packages/core/content-releases/server/src/services/release.ts

```diff
@@ -161,6 +164,89 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => ({
 
     return contentTypesMeta;
   },
+  async publish(releaseId: PublishRelease.Request['params']['id']) {
+    // We need to pass the type because entityService.findOne is not returning the correct type
+    const releaseWithPopulatedActionEntries = (await strapi.entityService.findOne(
+      RELEASE_MODEL_UID,
+      releaseId,
+      {
+        populate: {
+          actions: {
+            populate: {
+              entry: true,
+            },
+          },
+        },
+      }
+    )) as unknown as Release;
+
+    if (!releaseWithPopulatedActionEntries) {
+      throw new errors.NotFoundError(`No release found for id ${releaseId}`);
+    }
+
+    if (releaseWithPopulatedActionEntries.releasedAt) {
+      throw new errors.ValidationError('Release already published');
+    }
+
+    if (releaseWithPopulatedActionEntries.actions.length === 0) {
+      throw new errors.ValidationError('No entries to publish');
+    }
+
+    /**
+     * We separate publish and unpublish actions group by content type
+     */
+    const actions: {
+      [key: UID.ContentType]: {
+        publish: ReleaseAction['entry'][];
+        unpublish: ReleaseAction['entry'][];
+      };
+    } = {};
+    for (const action of releaseWithPopulatedActionEntries.actions) {
+      const contentTypeUid = action.contentType;
+
+      if (!actions[contentTypeUid]) {
+        actions[contentTypeUid] = {
+          publish: [],
+          unpublish: [],
+        };
+      }
+
+      if (action.type === 'publish') {
+        actions[contentTypeUid].publish.push(action.entry);
+      } else {
+        actions[contentTypeUid].unpublish.push(action.entry);
+      }
+    }
+
+    const entityManagerService = strapi.plugin('content-manager').service('entity-manager');
+
+    // Only publish the release if all action updates are applied successfully to their entry, otherwise leave everything as is
+    await strapi.db.transaction(async () => {
+      for (const contentTypeUid of Object.keys(actions)) {
+        const { publish, unpublish } = actions[contentTypeUid as UID.ContentType];
+
+        if (publish.length > 0) {
+          await entityManagerService.publishMany(publish, contentTypeUid);
+        }
+
+        if (unpublish.length > 0) {
+          await entityManagerService.unpublishMany(unpublish, contentTypeUid);
+        }
+      }
+    });
+
+    const release = await strapi.entityService.update(RELEASE_MODEL_UID, releaseId, {
```

Comment:
I would add a comment like:

```suggestion
		// When the transaction fails it throws an error, when it is successful proceed to updating the release
    const release = await strapi.entityService.update(RELEASE_MODEL_UID, releaseId, {
```

For someone not familiar with transactions (for example me 🙃 ) the code currently reads like no matter what update the release.